Data: refuse to register an already registered store

jsnajdr opened the pull request:
In https://github.com/Automattic/wp-calypso/pull/74346#issuecomment-1470328236 @noahtallen ran into an issue where a data store is registered twice. Because the offending store is defined in a JS module, and registered by a static top-level call:
```js
register( store );
```
and because this module is bundled in two webpack bundles, each for a different plugin.

In this PR I'm changing the `registerStoreInstance` behavior to refuse to register a second instance. At this moment I'm throwing an error, but that's just because I want unit and e2e tests to fail as badly as possible if they happen to re-register any store. In the final version there should be just a `return` statement.

There are concerns about backward compatibility, but my intuition is it should be a much smaller issue than it looks like. Does anyone really intentionally re-register a store? The only reason would be trying to clear it back to initial state.

Another argument against re-registering is that the `useSelect` hook wouldn't react to it anyway:
```js
useSelect( s => s( 'my-store' ).get(), [] );
```
won't see any update and won't trigger any rerender when the `my-store` is swapped for another instance.

what do you think @youknowriad?

jsnajdr committed:
Data: refuse to register an already registered store

tyxla commented on packages/data/src/registry.js:
```diff
@@ -215,6 +215,10 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	 */
 	function registerStoreInstance( name, store ) {
+		if ( stores[ name ] ) {
+			throw new Error( 'duplicate store oh no' );
```
Nit: Maybe we could improve the message a little bit.

jsnajdr replied:
It's going to be replaced by just `return;` eventually. The error is intentionally bad so that we don't forget about that 🙂

tyxla replied:
Sounds good.

We could still do a `console.error()` though. I think it's something worth informing developers about.

jsnajdr replied:
> We could still do a `console.error()` though.

That's a good idea, I added a `console.error` + `return` instead of throwing. It turns out that `registerBlockType` also does that.

On the other hand, `registerPlugin` prints a console error that plugin is already registered, but then proceeds to register the new version. Not sure if we really wanted that.

gziolo replied:
Here is an opportunity to define and align the behavior for all those APIs. I saw also @noahtallen commenting in https://github.com/WordPress/gutenberg/pull/49134#discussion_r1140688858 that we could use `warning` from the `@wordpress/warning` package.

jsnajdr replied:
Yes, I like the idea of using `warning` in these cases. Let's do that in a separate PR. I'd rather keep this one focused on the data package and its store registration code.

tyxla replied:
I think `@wordpress/warning` is a good idea 👍 And I'd be fine to see that done in a separate PR.

tyxla commented on packages/data/src/registry.js:
```diff
@@ -215,6 +215,10 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	 * @param {Object} store Store instance object (getSelectors, getActions, subscribe).
 	 */
 	function registerStoreInstance( name, store ) {
+		if ( stores[ name ] ) {
```
Such a change is worth a CHANGELOG entry.

jsnajdr replied:
Added ✅

tyxla reviewed (commented):
That change makes sense IMHO.

We are already `register()`ing stores in core, without checking if they've been registered already:

https://github.com/WordPress/gutenberg/blob/565af6c949a5cf817da3133091ea9cb85ff5f5ff/packages/core-data/src/index.js#L66

That being said, if `@wordpress/core-data` ends up in two separate bundles that run alongside the same app, this may create similar conflicts to the ones described here. That's undesired behavior and we need to protect against it.

noahtallen reviewed (commented):
I'm personally a big fan of this approach -- I think overwriting the store and anything in it isn't a good default behavior for duplicate registration :) 

Gutenberg packages mostly avoid this problem because they're almost always imported via dependency extraction. So this is an awesome change for people who use their own custom data stores 

The only bit of caution I have about this is the line [in the register function](https://github.com/WordPress/gutenberg/pull/49134/files#diff-ad03b0b03f26fb957ed38ffcf69efbce6ce88c06c16864569e80837e8b0ec5e8R277):

```js
registerStoreInstance( store.name, store.instantiate( registry ) );
``` 

Essentially, there would be two _different_ store objects (from the two different bundles importing the store), and `store.instantiate` would be called on each individually. Will this be a problem? (E.g. will we end up with two store copies with separate state?) It seemed like `store.instantiate` was doing a lot of work 🤔

jsnajdr committed:
Create store lazily, console error on duplicate store

jsnajdr commented:
> Essentially, there would be two _different_ store objects (from the two different bundles importing the store), and `store.instantiate` would be called on each individually. Will this be a problem?

Yes and no, because the second instance would be thrown away after detecting that the first instance is already registered. It would a mere inefficiency if there wasn't the persistence plugin that has some side effects. Hypothetically, it could delete the local storage after reading the state from it, or store data periodically. Then two instances are fighting each other.

I solved this in the `registerStoreInstance` function by making the store creation lazy. Instead of
```js
registerStoreInstance( name, instantiate() );
```
it's now
```js
registerStoreInstance( name, () => instantiate() );
```
and the instance will be created only after verifying that `stores[name]` is empty.

jsnajdr committed:
Add changelog entry

jsnajdr commented:
I addressed feedback and I believe the PR is in a mergeable state:
- made store instantiation lazy so that we don't create stores for nothing
- print a console error if a store is already registered
- added a changelog entry

One last thought: maybe we want a `registry.unregister` function, if someone really wants to register a new instance. Then they would need to unregister the old one first. But the downside is that we'd be opening a possibility to unregister someone else's stores, e.g., the core ones.

tyxla commented on packages/data/src/registry.js:
```diff
@@ -270,15 +280,17 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	function register( store ) {
-		registerStoreInstance( store.name, store.instantiate( registry ) );
+		registerStoreInstance( store.name, () =>
+			store.instantiate( registry )
```
Great idea for the lazy init 👍

tyxla commented on packages/data/src/registry.js:
```diff
@@ -211,10 +211,18 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	 */
-	function registerStoreInstance( name, store ) {
+	function registerStoreInstance( name, createStore ) {
+		if ( stores[ name ] ) {
```
I think this change is worth a unit test.

jsnajdr replied:
I agree, added a unit test that checks that after a second registration the old data are still there.

tyxla reviewed (approved):
LGTM 👍 

Great improvement, thanks @jsnajdr!

noahtallen commented on packages/data/src/registry.js:
```diff
@@ -270,15 +280,17 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	 * @param {StoreDescriptor} store Store descriptor.
 	 */
 	function register( store ) {
```
I possibly mis-understand this! The extra instance from the second bundle wouldn't ever be instantiated. If that's a problem, should we return the already registered store from `registerStoreInstance` in each of these functions? Then, at least the second store instance could get assigned to the one stored in the registry.

I guess that would make the consumer look more like `export const store = register( createReduxStore( ... ) )`. 

Happy to be corrected on this; I don't understand the internals of store registration that well :)

noahtallen commented on packages/data/src/registry.js:
```diff
@@ -211,10 +211,18 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
+	function registerStoreInstance( name, createStore ) {
+		if ( stores[ name ] ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Store "' + name + '" is already registered.' );
```
Should this be an info or warning call instead? IMO, we want this to be a supported option for stores

noahtallen commented on packages/data/src/registry.js:
```diff
@@ -270,15 +280,17 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 	 * @param {StoreDescriptor} store Store descriptor.
 	 */
 	function register( store ) {
```
I possibly mis-understand this! The extra instance from the second bundle wouldn't ever be instantiated. If that's a problem, should we also return the already registered store here? Then, at least the second store instance could get assigned to the one stored in the registry.

I guess that would make the consumer look more like `export const store = register( createReduxStore( ... ) )`. 

Happy to be corrected on this; I don't understand the internals of store registration that well :)

jsnajdr replied:
The `register` function currently doesn't return anything, and I'm keeping it that way. Would there be anything useful to return? The store descriptor, it's been passed as argument, the caller already has it. Not useful. And the actual instance of the store, it's a private property of the registry, you're supposed to access it solely through the `registry.select` and `registry.dispatch` functions.

The `@automattic/data-stores` package in Calypso uses a pattern with a `register` function, and you can implement it as follows:
```js
function register() {
  const plansStore = createReduxStore( STORE_NAME, ... );
  register( plansStore );
  return plansStore;
}
```
and then use as:
```js
const PLANS_STORE = Plans.register();

registry.select( PLANS_STORE ).getPlanProduct( ... );
```

The legacy `registerStore` function returns a `store` object, but that's all legacy. It's not even the store object returned by `storeDescription.instantiate` and used by the data registry, but the raw Redux store. It's useful for the `use` function for enabling plugins like `persistence`, nothing else.

noahtallen replied:
> you're supposed to access it solely through the registry.select and registry.dispatch functions.

Ah, now I understand :) Given this example:

```js
// store
export const fooStore = createReduxStore( ... );
register( fooStore );

// bundle1
import { fooStore as foo1 } from 'foo-store';
useSelect( select => select( foo1 ).selector() );

// bundle2
import { fooStore as foo2 } from 'foo-store';
useSelect( select => select( foo2 ).selector() );
```

If you were to somehow compare `foo1 === foo2`, they would be different objects. I was thinking this could be problematic, but I didn't realize that the important part of the store is just saved in the registry and not in the object returned from `createReduxStore` (and then exported for use.)

But realizing that `export const fooStore` is only relevant in that it lets the registry identify the store based on the store id, I see it won't be an issue.

youknowriad commented:
Sorry for the late response here. Yes, this change makes sense to me and I'm not really worried about he backward compatibility impact. I feel like this is more of a bug fix for folks registering the same store twice.

A dev note should be the way to go here.

aristath reviewed (approved):
LGTM 👍 
Great improvement, thank you @jsnajdr!

jsnajdr committed:
Add unit test to check double registration behavior
